### PR TITLE
Rewrite OpenXR hand tracking to reflect new architecture

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Work in progress, I'm rewriting the hand tracking manual page to reflect the new way things are structured in Godot 4.3.
About half way done.

Note that this manual page references https://github.com/godotengine/godot-demo-projects/pull/973 and that PR needs to be merged alongside this.

### CODE REVIEW

File: tutorials/xr/openxr_hand_tracking.rst [hunk 1]
```diff
@@ -1,164 +1,239 @@
...
+This can be done in the project settings:
+
+**add image of project settings**
+
+For some standalone XR devices you also need to configure the hand tracking extension in export settings,
+for instance for Meta Quest:
 
 .. image:: img/xr_enable_handtracking.webp
 
+**replace this image with an updated one**
```
Comment (BastiaanOlij) [inline tutorials/xr/openxr_hand_tracking.rst:88]: Reminder to self, need to do this.

File: tutorials/xr/openxr_body_tracking.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
+.. _doc_openxr_body_tracking:
+
+OpenXR body tracking
+=====================
```
Comment (AThousandShips) [inline tutorials/xr/openxr_body_tracking.rst:4]:
```suggestion
====================
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 3]
```diff
@@ -1,201 +1,355 @@
...
 
-Three, inferred hand tracking.
-This has been the de facto approach since the early days of VR.
-As we know the player is holding a controller and we know the position of this controller,
-we can infer where to render the players hand.
-Fingers can be positioned based on the controls the player is interacting with.
-Many modern VR controllers have additional sensors to help determine finger positions on the controller.
+When discussing hand tracking it is important to know that there are differences of opinion as to where lines are drawn.
+The practical result of this is that there are differences in implementation between the different OpenXR runtimes.
+You may find yourself in a place where chosen hardware doesn't support a piece of the puzzle or does things different
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:17]:
```suggestion
You may find yourself in a place where chosen hardware doesn't support a piece of the puzzle or does things differently
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 4]
```diff
@@ -1,201 +1,355 @@
...
+Recognising that there is an increasing demand for applications that can switch seamlessly
+between controller and hand tracking and the need some form of basic input capability,
+a number of extensions were added to the specification that provide some basic gesture recognision
+and can be used with the action map.
+
+The hand interaction profile
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `hand interaction profile extension <https://github.khronos.org/OpenXR-Inventory/extension_support.html#XR_EXT_hand_interaction>`_
+is a new core extension which supports pinch, grasp and poke gestures and related poses.
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:222]:
```suggestion
is a new core extension which supports pinch, grasp, and poke gestures and related poses.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 5]
```diff
@@ -1,201 +1,355 @@
...
-There is an extension that becomes available if HTC trackers are supported by the XR runtime.
-These are fully exposed through the action map system.
+This example simply logs the state for the left hand.
 
-Godot has full support for these and you can setup the trackers in the action map.
-Each tracker is assigned a usage within the SteamVR interface.
+If in this example no hand tracker is returned by ``get_tracker``,
+this means the hand tracking API is not supported on the XR runtime at all.
+
+If there is a tracker but `has_tracking_data` is false, the users hand is currently not being tracked.
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:193]:
```suggestion
If there is a tracker but `has_tracking_data` is false, the user's hand is currently not being tracked.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 6]
```diff
@@ -1,201 +1,355 @@
...
+
+When we look at the early days of VR the focus of the major platforms was on tracked controller based input.
+Here we are tracking a physical device that also has buttons for further input.
+From the tracking data we can infer the location of the players hands but no further information is known,
+traditionally it was left up to the game to implement a mechanism to display the players hand and animate
+the fingers based on further input from the controller, be it due to buttons being pressed or through proximity
+sensors.
+Often fingers are also placed based on context, what is the user holding, what is the action a user is performing.
+
+More recently optical hand tracking has become a popular solution, where cameras track the users hands
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:32]:
```suggestion
More recently optical hand tracking has become a popular solution, where cameras track the user's hands
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 7]
```diff
@@ -1,201 +1,355 @@
...
+or whether hand tracking is emulating a controller through this profile.
+
+.. image:: img/openxr_simple_controller_hand.webp
+
+XR runtimes are free to define how the simple controller profile operates,
+so there is also no certainty to how this profile is mapped to gestures.
+
+The most common mapping seems to be that ``select click`` is true
+when the tip of the thumb and index fingers are touching while the
+users palm is facing away from the user.
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:317]:
```suggestion
user's palm is facing away from the user.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 8]
```diff
@@ -1,201 +1,355 @@
...
+``menu click`` will be true when tip of the thumb and index fingers
+are touching while the users palm is facing towards the user.
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:319]:
```suggestion
are touching while the user's palm is facing towards the user.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 9]
```diff
@@ -1,201 +1,355 @@
...
+ *  A tracked node to position the hand.
+ *  A properly skinned hand mesh with skeleton.
+ *  A skeleton modifier that applies finger tracking data to the skeleton.
+
+.. image:: img/openxr_hand_tracking_nodes.webp
+
+Hand tracking node
+^^^^^^^^^^^^^^^^^^
+
+The hand tracking system uses separate hand trackers to track the position of the players hands
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:99]:
```suggestion
The hand tracking system uses separate hand trackers to track the position of the player's hands
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 10]
```diff
@@ -1,201 +1,355 @@
...
+enough from the other platforms that you need to do extra work.
+
+That said, recent improvements to the OpenXR specification are closing these gaps and as platforms implement
+these improvements we are getting closer to a future where we have either full portability between platforms
+or at least a clear way to detect the capabilities of a platform.
+
+When we look at the early days of VR the focus of the major platforms was on tracked controller based input.
+Here we are tracking a physical device that also has buttons for further input.
+From the tracking data we can infer the location of the players hands but no further information is known,
+traditionally it was left up to the game to implement a mechanism to display the players hand and animate
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:27]:
```suggestion
From the tracking data we can infer the location of the player's hands but no further information is known,
traditionally it was left up to the game to implement a mechanism to display the player's hand and animate
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 11]
```diff
@@ -1,201 +1,355 @@
...
+within our tracking space. 
+
+This information has been separated out for the following use cases:
+
+ *  Tracking happens in the local space of the :ref:`XROrigin3D <class_xrorigin3d>` node.
+    This node must be a child of the `XROrigin3D` node in order to be correctly placed.
+ *  This node can be used as an IK target when an upper body mesh with arms is used instead
+    of separate hand meshes.
+ *  Actual placement of the hands may be loosly bound to the tracking in scenarios such as
+    avatar creation UIs, fake mirrors or similar situations
```
Comment (AThousandShips) [inline tutorials/xr/openxr_hand_tracking.rst:109]:
```suggestion
    avatar creation UIs, fake mirrors, or similar situations
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 12]
```diff
@@ -1,201 +1,355 @@
...
+When we look at the early days of VR the focus of the major platforms was on tracked controller based input.
+Here we are tracking a physical device that also has buttons for further input.
+From the tracking data we can infer the location of the players hands but no further information is known,
+traditionally it was left up to the game to implement a mechanism to display the players hand and animate
+the fingers based on further input from the controller, be it due to buttons being pressed or through proximity
+sensors.
+Often fingers are also placed based on context, what is the user holding, what is the action a user is performing.
+
+More recently optical hand tracking has become a popular solution, where cameras track the users hands
+and full tracked data for the hand and finger position became available.
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:33]:
```suggestion
and full tracking data for the hand and finger positions becomes available.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 13]
```diff
@@ -1,201 +1,355 @@
...
+Many vendors saw this as completely separate from controller tracking and introduced independent APIs to
+access hand and finger position and orientation data.
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:35]:
```suggestion
access hand and finger positions and orientation data.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 14]
```diff
@@ -1,201 +1,355 @@
...
+When handling input it was up to the game developer to implement a gesture detection mechanism.
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:36]:
```suggestion
When handling input, it was up to the game developer to implement a gesture detection mechanism.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 15]
```diff
@@ -1,201 +1,355 @@
...
+the XR runtime/hardware or that add further functionality accross this divide.
+The problem that currently does remain is that there are gaps in adopting these extensions,
+with some platforms thus not reporting capabilities to their full extent.
+As such you may need to test for the features available on specific hardware
+and adjust your approach accordingly.
+
+Demo project
+------------
+
+The information presented on this page was used to create a demo projection that can be found
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:61]:
```suggestion
The information presented on this page was used to create a demo project that can be found
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 16]
```diff
@@ -1,201 +1,355 @@
...
+---------------------
+
+As mentioned in our introduction, the hand tracking API is primarily used with optical hand tracking
+and on many platforms only works when the user is not holding a controller.
+Some platforms support controller inferred hand tracking meaning that you will get hand tracking data
+even if the user is holding a controller.
+This includes SteamVR, Meta Quest (currently native only but Meta link support is likely coming)
+and hopefully soon others as well.
+
+The hand tracking implementation in Godot has been standardised around the Godot Humanoid Skeleton
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:75]: Probably not important, but I think we try to use US (vs UK) spellings:

```suggestion
The hand tracking implementation in Godot has been standardized around the Godot Humanoid Skeleton
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 17]
```diff
@@ -1,201 +1,355 @@
...
+
+This information has been separated out for the following use cases:
+
+ *  Tracking happens in the local space of the :ref:`XROrigin3D <class_xrorigin3d>` node.
+    This node must be a child of the `XROrigin3D` node in order to be correctly placed.
+ *  This node can be used as an IK target when an upper body mesh with arms is used instead
+    of separate hand meshes.
+ *  Actual placement of the hands may be loosly bound to the tracking in scenarios such as
+    avatar creation UIs, fake mirrors or similar situations
+    resulting in the hand mesh and finger tracking being localised elsewhere.
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:110]: Another US vs UK spelling:

```suggestion
    resulting in the hand mesh and finger tracking being localized elsewhere.
```

File: tutorials/xr/openxr_hand_tracking.rst [hunk 18]
```diff
@@ -1,201 +1,355 @@
...
+if controllers are used.
+In the action map you can map various inputs like the trigger or joystick on the controller
+to an action. This can then drive logic in your game.
+
+When hand tracking is used we originally had no such inputs,
+inputs are driven by gestures made by the user such as making a fist to grab
+or pinching the thumb and index finger together to select something.
+It was up to the game developer to implement this.
+
+Recognising that there is an increasing demand for applications that can switch seamlessly
```
Comment (dsnopek) [inline tutorials/xr/openxr_hand_tracking.rst:213]: US vs UK spelling:

```suggestion
Recognizing that there is an increasing demand for applications that can switch seamlessly
```

**Review by dsnopek — COMMENTED** [review]
Thanks! This looks great!

I only have a couple language nit-picks.

**Comment by BastiaanOlij** [comment]
I think I got all the reported grammar issue, thanks for the proof reading @AThousandShips and @dsnopek

**Review by dsnopek — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Review by skyace65 — CHANGES_REQUESTED** [review]

**Review by skyace65 — APPROVED** [review]

**Review by mhilbrunner — APPROVED** [review]

**Review by m4gr3d — APPROVED** [review]

**Comment by BastiaanOlij** [comment]
Thanks for the approvals everyone, please also check out the demo project that accompanies this.

**Comment by skyace65** [comment]
I can check that the demo runs on my computer later this week. Unfortunately though my headset doesn't have hand tracking.

**Comment by BastiaanOlij** [comment]
> I can check that the demo runs on my computer later this week. Unfortunately though my headset doesn't have hand tracking.

If you have steamVR it'll work on inferred hand tracking, all should still work :)

**Comment by mhilbrunner** [comment]
This PR is good to go and just waiting on the demo to be merged. Unfortunately, I currently lack the bandwith to check it out myself. Please poke me once that PR is merged so we can merge this one :)

**Comment by skyace65** [comment]
XR Demo is merged as well. Thanks for your work on both of these!

**Pull request merged**